### PR TITLE
allow zero and null as override values

### PR DESCRIPTION
**Pull request by hidalgofdz**
Add the possibility to pass 0 or null as an override value. I'm new with typescript so I'm not sure if I'am using types correctly.

### CODE REVIEW

File: src/index.test.ts [hunk 1]
```diff
@@ -122,6 +122,40 @@ describe('test-data-bot', () => {
...
+
+    const product = productBuilder({ overrides: { amount: 0 } });
+    expect(product).toEqual({
+      amount: 0,
+    });
+  });
+
+  it('lets a value be overridden with null when building an instance', () => {
+    interface User {
+      name?: string | null;
```
Comment (kimroen) [inline src/index.test.ts:144]: This question mark shouldn't be necessary.

File: src/index.ts [hunk 2]
```diff
@@ -84,6 +84,23 @@ const buildTimeTraitsArray = <FactoryResultType>(
   return Array.isArray(traits) ? traits : [traits];
 };
 
+const getValueOrOverride = (
+  overrides: Overrides<any>,
+  traitOverrides: Overrides<any>,
+  fieldValue: any,
+  fieldKey: string
+): any => {
```
Comment (kimroen) [inline src/index.ts:92]: We don't want the `any` types here, we need to make this function have accurate types.

Comment (hidalgofdz) [reply src/index.ts:92]: What do you think about something like this?

```
const getValueOrOverride = <FactoryResultType> (
  overrides: Overrides<FactoryResultType>,
  traitOverrides: Overrides<FactoryResultType>,
  fieldValue: Field,
  fieldKey: string
): Field => {
```

File: src/index.test.ts (hunk 1 above)
Comment (hidalgofdz) [reply src/index.test.ts]: I remove the question mark.

File: src/index.ts (hunk 2 above)
Comment (kimroen) [reply src/index.ts]: Yeah, I think that looks reasonable.

**Comment by nmccrea** [comment]
Just ran into this issue. My workaround for now is to use `perBuild()` to set the falsy value:

```javascript
const thingBuilder = build({
  fields: {
    truthyOrFalsyField: null,
  },
  traits: {
    truthy: {
      overrides: { truthyOrFalsyField: 1 },
    },
    falsy: {
      overrides: { truthyOrFalsyField: perBuild(() => 0) },
    },
  },
)
```

Regarding the discussion in #287, I think it should be possible to override any field with any value, including `undefined`. I don't think the library should make any assumptions about what are and are not appropriate override values.

I think the conditionals should only check if the field key exists in the list of override keys, and not be based on the override _value_ at all:

```typescript
const getValueOrOverride = <FactoryResultType> (
  overrides: Overrides<FactoryResultType>,
  traitOverrides: Overrides<FactoryResultType>,
  fieldValue: Field,
  fieldKey: string
): Field => {
  if (Object.keys(overrides).includes(fieldKey)) {
    return overrides[fieldKey]
  }

  if (Object.keys(traitOverrides).includes(fieldKey)) {
    return traitOverrides[fieldKey];
  }

  return fieldValue;
};
```

Also thanks so much for this library!

**Comment by jackfranklin** [comment]
Merging this but will follow-up with a patch based on @nmccrea's code as I like that suggestion. Thanks everyone!

**Pull request merged**